### PR TITLE
SONAR-8061 speed up issue WebServices by removing duplicate SQL statements during WS call

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java [hunk 1]
```diff
@@ -87,9 +96,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
...
+      } else {
+        List<UserDto> loadedUsers = dbClient.userDao().selectByLogins(dbSession, loginsToLoad);
+        result.setUsers(concat(preloadedUsers.stream(), loadedUsers.stream()).collect(toList(preloadedUsers.size() + loadedUsers.size())));
+      }
+    }
+  }
+
+  private void loadComponents(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> preloadedComponents = preloadedResponseData.getComponents();
+    Set<String> preloadedComponentUuids = preloadedComponents.stream().map(ComponentDto::uuid).collect(MoreCollectors.toSet(preloadedComponents.size()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java]: MAJOR A "NullPointerException" could be thrown; "preloadedComponents" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java [hunk 2]
```diff
@@ -87,9 +96,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
...
+    result.addComponents(dbClient.componentDao().selectSubProjectsByComponentUuids(dbSession, collector.getComponentUuids()));
+    addProjectUuids(collector, dbSession, result);
+  }
+
+  private void addProjectUuids(Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> loadedComponents = result.getComponents();
+    for (ComponentDto component : loadedComponents) {
+      collector.addProjectUuid(component.projectUuid());
+    }
+    Set<String> loadedProjectUuids = loadedComponents.stream().filter(cpt -> cpt.uuid().equals(cpt.projectUuid())).map(ComponentDto::uuid).collect(MoreCollectors.toSet());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java]: MAJOR A "NullPointerException" could be thrown; "loadedComponents" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java [hunk 3]
```diff
@@ -44,21 +48,44 @@ public IssueUpdater(DbClient dbClient, IssueStorage issueStorage, NotificationMa
...
   public void saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment) {
-    issueStorage.save(session, issue);
     Optional<RuleDefinitionDto> rule = getRuleByKey(session, issue.getRuleKey());
     ComponentDto project = dbClient.componentDao().selectOrFailByUuid(session, issue.projectUuid());
+    ComponentDto component = dbClient.componentDao().selectOrFailByUuid(session, issue.componentUuid());
+    saveIssue(session, issue, context, comment, rule, project, component);
+  }
+
+  private void saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment,
+    Optional<RuleDefinitionDto> rule, ComponentDto project, ComponentDto component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java]: MINOR Specify a "RuleDefinitionDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java [hunk 4]
```diff
@@ -44,21 +49,46 @@ public IssueUpdater(DbClient dbClient, IssueStorage issueStorage, NotificationMa
...
+
+  public IssueDto saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment) {
+    Optional<RuleDefinitionDto> rule = getRuleByKey(session, issue.getRuleKey());
+    ComponentDto project = dbClient.componentDao().selectOrFailByUuid(session, issue.projectUuid());
+    ComponentDto component = dbClient.componentDao().selectOrFailByUuid(session, issue.componentUuid());
+    return saveIssue(session, issue, context, comment, rule, project, component);
+  }
+
+  private IssueDto saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment,
+    Optional<RuleDefinitionDto> rule, ComponentDto project, ComponentDto component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java]: MINOR Specify a "RuleDefinitionDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/AssignAction.java [hunk 5]
```diff
@@ -121,8 +120,9 @@ private void assign(String issueKey, @Nullable String assignee) {
       }
       IssueChangeContext context = IssueChangeContext.createUser(new Date(system2.now()), userSession.getLogin());
       if (issueFieldsSetter.assign(issue, user, context)) {
-        issueUpdater.saveIssue(dbSession, issue, context, null);
+        return issueUpdater.saveIssueAndPreloadSearchResponseData(dbSession, issue, context, null);
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AssignAction.java]: Note that the user could also be part of the preloaded response. It implies to change IssueChangeContext by replacing user login by UserDto.
This improvement does not sound to be a must-have.

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AssignAction.java]: almost true.

The login in IssueChangeContext is the login of the caller of the WS.
The users which are loaded from DB to build the response is the assignee of the issue. Apparently, we don't care about the current user, except for issues/search (see https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java#L415-L419).
In addition, the request on table USERS by id is fast (table is small compared to projects or issues).

In the current state, the next winner is removing the SQL request on table PROJECTS which retrieves all modules. This request is commented as "here for backward compatibility" but it is costly (50ms locally) (this one: https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java#L123).

**Pull request merged**